### ISSUE TITLE
OneCycleLR does not follow the algorithm described by the original paper

### ISSUE THREAD

**Issue by anibali** [pytorch/pytorch#40362]
## 🐛 Bug

`torch.optim.lr_scheduler.OneCycleLR` claims to be an implementation of the schedule originally described in the paper [Super-Convergence: Very Fast Training of Neural Networks Using Large Learning Rates](https://arxiv.org/abs/1708.07120), but does not seem to match the algorithm described by the authors.

Here is a quote from that paper:

> Here we suggest a slight modification of cyclical learning rate policy for super-convergence; always use one cycle that is smaller than the total number of iterations/epochs and allow the learning rate to decrease several orders of magnitude less than the initial learning rate for the remaining iterations. We named this learning rate policy “1cycle” and in our experiments this policy allows an improvement in the accuracy.

This implies a learning rate schedule with 3 phases:

1. Increase from initial learning rate to max learning rate
2. Decrease from max learning rate to initial learning rate
3. Decrease from initial learning rate to minimum learning rate

![expected](https://user-images.githubusercontent.com/226707/85246285-92880e80-b48d-11ea-9874-76f76949f9ac.png)

However, in its current form `OneCycleLR` only has two phases:

1. Increase from initial learning rate to max learning rate
2. Decrease from max learning rate to minimum learning rate

![actual](https://user-images.githubusercontent.com/226707/85246291-99168600-b48d-11ea-97f3-41d787041ca1.png)

There is a similar issue with momentum only having 2 phases instead of 3.

## To Reproduce

```python
from torch import nn
from torch import optim
from torch.optim.lr_scheduler import OneCycleLR
import matplotlib.pyplot as plt

model = nn.Linear(1, 1)
max_lr = 1.0
momentum = 0.9
epochs = 150
batches_per_epoch = 1000
optimiser = optim.SGD(model.parameters(), lr=max_lr, momentum=momentum)
scheduler = OneCycleLR(optimiser, max_lr, epochs=epochs, steps_per_epoch=batches_per_epoch,
                       pct_start=0.475, anneal_strategy='linear', cycle_momentum=True,
                       base_momentum=0.85, max_momentum=momentum, div_factor=10,
                       final_div_factor=1e4)

ys = []
for _ in range(epochs):
    for _ in range(batches_per_epoch):
        ys.append(optimiser.param_groups[0]['lr'])
        scheduler.step()
plt.title('OneCycleLR schedule')
plt.ylabel('Learning rate')
plt.xlabel('Step')
plt.plot(ys, c='red')
plt.show()
```

## Expected behavior

The plot should look like the first figure in this report ("Expected schedule"), but instead looks like the second ("OneCycleLR schedule").

## Environment

PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Ubuntu 20.04 LTS
GCC version: (Ubuntu 9.3.0-10ubuntu2) 9.3.0
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 980
Nvidia driver version: 440.64
cuDNN version: Could not collect

## Additional context

There are other 1cycle implementations/write-ups around the Internet which correctly identify that the schedule has three phases. Some examples are:

* https://sgugger.github.io/the-1cycle-policy.html
* https://github.com/titu1994/keras-one-cycle

cc @ezyang @gchanan @zou3519 @vincentqb

**Comment by zou3519** [comment]
I don't have context behind the paper but I am labeling this as tentatively high-pri for investigation because there appears to be a silent correctness issue.

**Comment by vincentqb** [comment]
Thank you, that's a good observation! The authors do label explicitly "1cycle" as the three phase you describe, but they also say this is a "slight modification" that works better in their experiment. In any case, it would be a good idea to have that feature in. @anibali  would you like to open a pull request for this? Could you share the code to generate the plots and learning rates above? They could be used as tests.

In the meanwhile, adding to the documentation would clarify the implementation, so let's highlighting the fact that the implementation is of only the first two phases in the docstring. @anibali do you want to do this?

**Comment by vincentqb** [comment]
The code only reproduces the second plot. Do you have the code for the first plot?

**Comment by anibali** [comment]
@vincentqb I would contend that the current implementation is still incorrect (or at the very least very misleading) if it is intended to be for the first two phases, since the first two phases should be symmetrical. However, in the current PyTorch implementation the parameter `final_div_factor` is being applied to phase 2 (when it should only be relevant for the currently missing phase 3). It also has a pretty strange interaction with `pct_start`, which should (in the symmetrical two-phase case) always be 0.5, but currently has a default value of 0.3. So to get a reasonable 2-phase setup happening, the user must explicitly set `pct_start=0.5` and `final_div_factor=1`, and it's much too easy to end up with a schedule that does not mirror anything proposed in the paper.

I'd be happy to work on a PR for adding in phase 3 and providing better defaults and safety rails, but I think that it will necessarily need to change the behaviour of the code to be more in-line with what you'd expect after reading the paper. Is it OK to make changes that are not backwards compatible?

> The code only reproduces the second plot. Do you have the code for the first plot?

The second plot is based on my own implementation from before `OneCycleLR` existed. I haven't fully investigated the correctness of the boundaries (eg. it might be shifted by 1 step or something), so I'm hesitant to put it out there as the gold standard.

**Comment by vincentqb** [comment]
In order to resolve this, we'd need a test case that fails with the current implementation with the exact expected values. In particular, in #21258 there are tests that were added. Is one of them not agreeing with an external references? We should also verify that the parameters available indeed can't reproduce that source.

**Comment by anibali** [comment]
I feel like perhaps you haven't understood what I am trying to say. The current implementation of 1cycle in PyTorch does not allow you to replicate the recommended policy from the paper (3 phases, first two being symmetrical) and defaults to a policy that I've never seen suggested anywhere (asymmetrical 2 phases, no third phase). The best you can achieve (in terms of matching the source paper) is tweaking the parameters to achieve a symmetrical 2 phase policy, but this is not immediately obvious. The end result will be a lot of people thinking that they are using 1cycle, when in fact they are using some other schedule which is somewhat similar.

What is more, the parameters are misleading since `final_div_factor` sounds like it refers to the third phase described in the paper, when in fact it controls asymmetry in the second phase (which, again, is not really a thing). Additionally, `pct_start` sounds like it controls the length of the first two phases, but once again it controls asymmetry.

In terms of the tests, I can interpret what I think the parameters `final_div_factor` and `pct_start` *should* control in the context of the recommended 3-phase version of 1cycle, and calculate expected learning rate values based on those assumptions. Is that what you are after?

**Comment by vincentqb** [comment]
Thanks for the reply :) To paraphrase what you say, there is two issues: (1) current implementation gives 2 asymetrical phases, but the paper presents 2 symmetrical + a third, (2) `final_div_factor` and `pct_start` may be confusing names. Is that what you meant?

I'm ok with updating the the implementation. Ideally, we will avoid BC-breaking changes, and add to the documentation how to align with the original paper.

What I meant to say is just that: to be sure that we are getting it right, we should have some reference to compare against. Thoughts?

**Comment by anibali** [comment]
Yes, I think that we are pretty much on the same page now. To elaborate on (2), those parameters currently control the schedule in strange ways which are not suggested in the source paper, which I think could be a major source of confusion in the future (especially since the default values of the parameters results in a bizarre schedule).

This means that we probably have to choose between breaking backwards compatibility or keeping a counter-intuitive API which makes it way too easy to do the wrong thing. Since asymmetrical 2 phases isn't really an established thing, it should either be in a separate scheduler or require deliberate opting in from the user (and definitely not be the default behaviour in my opinion).

I agree that having reference LR values in a test case for the full 3-phase version of 1cycle would be useful, I can work on that when I have some time.

**Comment by vincentqb** [comment]
Awesome, thanks a lot for the input, and please feel free to open a pull request for this :)

**Comment by MANU-CHAUHAN** [comment]
Hi all,

I tried using OneCycleLR and changed `pct_start` value and observed LR at start of every epoch (after doing scheduler.step at every batch), but LR does not seem to increase and decrease as per `pct_start` float value given, it resorts to default of 0.3. 

Am I doing something wrong ??
```
max_lr = 5.61e-2

optimizer = optim.SGD(model.parameters(), lr = 5.61e-3, nesterov=False)

scheduler = torch.optim.lr_scheduler.OneCycleLR(optimizer=optimizer,
                                                max_lr=max_lr,
                                                pct_start=0.16, # 24 * 0.16 = 3.84, so 3 epochs will be used to increase LR
                                                anneal_strategy='linear',
                                                verbose=False, epochs=24,
                                                steps_per_epoch=len(train_loader),
                                                base_momentum=0.87,
                                                max_momentum=0.9)
```


LR at epoch 1 start: --> [0.002244]

LR at epoch 2 start: --> [0.01630636811254396]

LR at epoch 3 start: --> [0.030368736225087917]

LR at epoch 4 start: --> [0.04443110433763189]

LR at epoch 5 start: --> [0.055626368512609325]

LR at epoch 6 start: --> [0.05284364154832361]

LR at epoch 7 start: --> [0.0500609145840379]

LR at epoch 8 start: --> [0.04727818761975218]

LR at epoch 9 start: --> [0.04449546065546647]


If `pct_start=0.16` shouldn't that be `floor(epochs * pct_start)` ? So 24 * 0.16 = 3.84 i.e. first 3 epochs to increase LR and reduce after that ? Shall I open a separate issue for this?
